### ISSUE TITLE
question about miou

### ISSUE THREAD

**Issue by xuchen-dev** [CoinCheung/BiSeNet#154]
i use my own dataset to train  13 classes  
i run the demo.py  to inference ，the result is look good like this
![image](https://user-images.githubusercontent.com/79962968/120635614-41570c00-c49f-11eb-876e-26b92663403e.png)
but  the miou is very low  just 0.19
about the evaluation part  i just change the  classes  19 --> 13
and because my picture is 1280*720  i change the 
class TransformationVal(object):
    def __call__(self, im_lb):
        im, lb = im_lb['im'], im_lb['lb']
        im = cv2.resize(im, (2048, 1024))
        lb = cv2.resize(lb, (2048, 1024), interpolation=cv2.INTER_NEAREST)
        return dict(im=im, lb=lb)

beg for you help ,if you have time ,i will very grateful

**Comment by CoinCheung** [comment]
Hi, 
What is your training crop size, and what is your image size in `demo.py` ?
Would you tell me how did you obtain 0.19 miou, with multi-scale evaluation or crop evaluation ?

**Comment by xuchen-dev** [comment]
|   single_scale |   single_scale_crop |   ms_flip |   ms_flip_crop |
|----------------+---------------------+-----------+----------------|
|       0.192733 |            0.191332 |  0.193444 |       0.191755 |

train  cropsize = [512, 1024]        in the config didn't change

i have try   

class TransformationVal(object):
def call(self, im_lb):
im, lb = im_lb['im'], im_lb['lb']
im = cv2.resize(im, (512, 1024))
lb = cv2.resize(lb, (512, 1024), interpolation=cv2.INTER_NEAREST)
return dict(im=im, lb=lb)

also  try resize 1024,512 
the result is also bad   

img size in demo.py is  1280*720  
because the size      in demo.py  i add       im = cv2.resize(im, (1024, 512))  change it to 1024*512
i ues the 13classes gt draw this picture   white  represent 255
![image](https://user-images.githubusercontent.com/79962968/120764495-c814f300-c54a-11eb-8eee-f4b7ab5afadb.png)

**Comment by xuchen-dev** [comment]
the net  need  /32        720 can not /32   will raise error

**Comment by CoinCheung** [comment]
Maybe you can try 1280x736 or 1280x704 which is closer to original size. Also please do not resize your label file since this would change ground truth, maybe you can first resize you input tensor to (736x1280) and resize the output back to (720x1280). Hope this would work.

**Comment by CoinCheung** [comment]
I think you do not need TransformationVal, since I have done the resize part for you. See [here]( https://github.com/CoinCheung/BiSeNet/blob/90de8f9bdb6281f700087a5384996b18a4c6dae7/tools/evaluate.py#L56)

**Comment by xuchen-dev** [comment]
yes  this line : 

  logits = F.interpolate(logits, size=size,   mode='bilinear', align_corners=True)

change the result to label shape 

i follow you suggestion change the img to 736*1280  and do not size the label   1280*736 too   it dosent work :
if there  something in miou calculation need change i didn't aware

**Comment by CoinCheung** [comment]
I do not think so, since I also used this code to evaluate on other datasets. I did not observe this problem. Would you please let me know how did you changed the code to fit for your dataset ?

**Comment by xuchen-dev** [comment]
**in train.py**

net = model_factory[cfg.model_type](13)   #  19 -->13

**in basedataset.py**    
i change the way to get data   i have printed this is no wrong 

class BaseDataset(Dataset):
    '''
    '''
    def __init__(self, dataroot, trans_func=None, mode='train'):
        super(BaseDataset, self).__init__()
        assert mode in ('train', 'val', 'test')
        self.mode = mode
        self.trans_func = trans_func

        self.lb_map = None
        if mode == 'train':
            image_list = os.listdir(dataroot + str('train/data'))
            label_list = [img_path[:-4] + '.png' for img_path in image_list]

            img_path = dataroot + 'train/data/'
            label_path = dataroot + 'train/label/'

            self.img_paths = [os.path.join(img_path, name) for name in image_list]
            self.lb_paths = [os.path.join(label_path, name) for name in label_list]

        if mode == 'val':
            image_list = os.listdir(dataroot + str('val/data'))
            label_list = [img_path[:-4] + '_render.png' for img_path in image_list]

            img_path = dataroot + 'val/data/'
            label_path = dataroot + 'val/label/'

            self.img_paths = [os.path.join(img_path, name) for name in image_list]
            self.lb_paths = [os.path.join(label_path, name) for name in label_list]

        assert len(self.img_paths) == len(self.lb_paths)
        self.len = len(self.img_paths)
        #print('################')
        #print(self.len)

def __getitem__(self, idx):
        impth, lbpth = self.img_paths[idx], self.lb_paths[idx]
        #print(impth,lbpth)
        #print('############')
        img, label = cv2.imread(impth)[:, :, ::-1], cv2.imread(lbpth, 0)
        #print(img.shape,label.shape)
        if not self.lb_map is None:
            label = self.lb_map[label]
        im_lb = dict(im=img, lb=label)

        if not self.trans_func is None:
            im_lb = self.trans_func(im_lb)
        im_lb = self.to_tensor(im_lb)
        img, label = im_lb['im'], im_lb['lb']

        #print(img.shape,label.shape)
        a = np.unique(label)
        #print(a,img.shape,max(a))
        return img.detach(), label.unsqueeze(0).detach()

class TransformationVal(object):

    def __call__(self, im_lb):
        im, lb = im_lb['im'], im_lb['lb']

        im = cv2.resize(im, (736,1280))
        #lb = cv2.resize(lb, (1024,512), interpolation=cv2.INTER_NEAREST)

        return dict(im=im, lb=lb)

**in city_cv2.py:**
labels_info = [
    {"hasInstances": False, "category": "void", "catid": 0, "name": "unlabeled", "ignoreInEval": True, "id": 0, "color": [0, 0, 0], "trainId": 12},
    {"hasInstances": False, "category": "void", "catid": 0, "name": "ego vehicle", "ignoreInEval": True, "id": 1, "color": [0, 0, 0], "trainId": 1},
    {"hasInstances": False, "category": "void", "catid": 0, "name": "rectification border", "ignoreInEval": True, "id": 2, "color": [0, 0, 0], "trainId": 1},
    {"hasInstances": False, "category": "void", "catid": 0, "name": "out of roi", "ignoreInEval": True, "id": 3, "color": [0, 0, 0], "trainId": 1},
    {"hasInstances": False, "category": "void", "catid": 0, "name": "static", "ignoreInEval": True, "id": 4, "color": [0, 0, 0], "trainId": 3},
    {"hasInstances": False, "category": "void", "catid": 0, "name": "dynamic", "ignoreInEval": True, "id": 5, "color": [111, 74, 0], "trainId": 3},
    {"hasInstances": False, "category": "void", "catid": 0, "name": "ground", "ignoreInEval": True, "id": 6, "color": [81, 0, 81], "trainId": 1},
    {"hasInstances": False, "category": "flat", "catid": 1, "name": "road", "ignoreInEval": False, "id": 7, "color": [128, 64, 128], "trainId": 8},
    {"hasInstances": False, "category": "flat", "catid": 1, "name": "sidewalk", "ignoreInEval": False, "id": 8, "color": [244, 35, 232], "trainId": 4},
    {"hasInstances": False, "category": "flat", "catid": 1, "name": "parking", "ignoreInEval": True, "id": 9, "color": [250, 170, 160], "trainId": 0},
    {"hasInstances": False, "category": "flat", "catid": 1, "name": "rail track", "ignoreInEval": True, "id": 10, "color": [230, 150, 140], "trainId": 1},
    {"hasInstances": False, "category": "construction", "catid": 2, "name": "building", "ignoreInEval": False, "id": 11, "color": [70, 70, 70], "trainId": 255},
    {"hasInstances": False, "category": "construction", "catid": 2, "name": "wall", "ignoreInEval": False, "id": 12, "color": [102, 102, 156], "trainId": 3},
    {"hasInstances": False, "category": "construction", "catid": 2, "name": "fence", "ignoreInEval": False, "id": 13, "color": [190, 153, 153], "trainId": 10},
    {"hasInstances": False, "category": "construction", "catid": 2, "name": "guard rail", "ignoreInEval": True, "id": 14, "color": [180, 165, 180], "trainId": 255},
    {"hasInstances": False, "category": "construction", "catid": 2, "name": "bridge", "ignoreInEval": True, "id": 15, "color": [150, 100, 100], "trainId": 1},
    {"hasInstances": False, "category": "construction", "catid": 2, "name": "tunnel", "ignoreInEval": True, "id": 16, "color": [150, 120, 90], "trainId": 0},
    {"hasInstances": False, "category": "object", "catid": 3, "name": "pole", "ignoreInEval": False, "id": 17, "color": [153, 153, 153], "trainId": 9},
    {"hasInstances": False, "category": "object", "catid": 3, "name": "polegroup", "ignoreInEval": True, "id": 18, "color": [153, 153, 153], "trainId": 10},
    {"hasInstances": False, "category": "object", "catid": 3, "name": "traffic light", "ignoreInEval": False, "id": 19, "color": [250, 170, 30], "trainId": 11},
    {"hasInstances": False, "category": "object", "catid": 3, "name": "traffic sign", "ignoreInEval": False, "id": 20, "color": [220, 220, 0], "trainId": 3},
    {"hasInstances": False, "category": "nature", "catid": 4, "name": "vegetation", "ignoreInEval": False, "id": 21, "color": [107, 142, 35], "trainId": 1},
    {"hasInstances": False, "category": "nature", "catid": 4, "name": "terrain", "ignoreInEval": False, "id": 22, "color": [152, 251, 152], "trainId": 255},
    {"hasInstances": False, "category": "sky", "catid": 5, "name": "sky", "ignoreInEval": False, "id": 23, "color": [70, 130, 180], "trainId": 7},
    {"hasInstances": True, "category": "human", "catid": 6, "name": "person", "ignoreInEval": False, "id": 24, "color": [220, 20, 60], "trainId": 10},
    {"hasInstances": True, "category": "human", "catid": 6, "name": "rider", "ignoreInEval": False, "id": 25, "color": [255, 0, 0], "trainId": 10},
    {"hasInstances": True, "category": "vehicle", "catid": 7, "name": "car", "ignoreInEval": False, "id": 26, "color": [0, 0, 142], "trainId": 11},
    {"hasInstances": True, "category": "vehicle", "catid": 7, "name": "truck", "ignoreInEval": False, "id": 27, "color": [0, 0, 70], "trainId": 5},
    {"hasInstances": True, "category": "vehicle", "catid": 7, "name": "bus", "ignoreInEval": False, "id": 28, "color": [0, 60, 100], "trainId": 1},
    {"hasInstances": True, "category": "vehicle", "catid": 7, "name": "caravan", "ignoreInEval": True, "id": 29, "color": [0, 0, 90], "trainId": 2},
    {"hasInstances": True, "category": "vehicle", "catid": 7, "name": "trailer", "ignoreInEval": True, "id": 30, "color": [0, 0, 110], "trainId": 1},
    {"hasInstances": True, "category": "vehicle", "catid": 7, "name": "train", "ignoreInEval": False, "id": 31, "color": [0, 80, 100], "trainId": 1},
    {"hasInstances": True, "category": "vehicle", "catid": 7, "name": "motorcycle", "ignoreInEval": False, "id": 32, "color": [0, 0, 230], "trainId": 10},
    {"hasInstances": True, "category": "vehicle", "catid": 7, "name": "bicycle", "ignoreInEval": False, "id": 33, "color": [119, 11, 32], "trainId": 2},
    {"hasInstances": False, "category": "vehicle", "catid": 7, "name": "license plate", "ignoreInEval": True, "id": 34, "color": [0, 0, 142], "trainId": 4},
    {"hasInstances": False, "category": "vehicle", "catid": 7, "name": "license plate", "ignoreInEval": True, "id": 35, "color": [0, 0, 142], "trainId": 10},
    {"hasInstances": False, "category": "vehicle", "catid": 7, "name": "license plate", "ignoreInEval": True, "id": 36, "color": [0, 0, 142], "trainId": 6},
    {"hasInstances": False, "category": "vehicle", "catid": 7, "name": "license plate", "ignoreInEval": True, "id": 37, "color": [0, 0, 142], "trainId": 3},
    {"hasInstances": False, "category": "vehicle", "catid": 7, "name": "license plate", "ignoreInEval": True, "id": 38, "color": [0, 0, 142], "trainId": 10}
]

  self.n_cats = 13


**this i didn't change**  
 self.to_tensor = T.ToTensor(
            mean=(0.3257, 0.3690, 0.3223), # city, rgb
            std=(0.2112, 0.2148, 0.2115),
        )

**Comment by CoinCheung** [comment]
Does your dataset have label mapping ? Do you really need this: https://github.com/CoinCheung/BiSeNet/blob/90de8f9bdb6281f700087a5384996b18a4c6dae7/lib/cityscapes_cv2.py#L68 ?

**Comment by xuchen-dev** [comment]
yes  my data is 39 classes   map  to  13 classes

**Comment by CoinCheung** [comment]
Sorry that I cannot find out the problem, maybe the result is correct. Since you are computing miou which can be dragged down by certain categorical ious(we compute average iou of each category and compute their mean value. If some categories have low iou, the final result will be low). 

If you find errors in the code, please do not hesitate to tell me. I will fix them.

**Comment by Pyten** [comment]
Check the shape setting of TransformationVal in cityscapes_cv2.py, figure out if the width and height are confused.

**Comment by xuchen-dev** [comment]
i have try some shapes  736x1280（1280x736 ） 1280x704 704x1280   512x1024 1024x512  ，

**Comment by CoinCheung** [comment]
Hi, I am not sure whether this is still a problem now. 
Since you are using your own dataset, it is possible that your dataset is not balanced. The major categories have higher miou, which let you observe good result, and the minor categories have lower miou which drag down global miou. 
This is merely my personal guess. If you want to verify, you can compute fw_mou, and see if the gap between miou and fw_miou is large.

**Comment by CoinCheung** [comment]
I am closing this, since much time has past, and maybe you have figured out correct explanations.
